b2sdk/transfer/inbound/downloader/parallel.py

```diff
@@ -294,67 +292,74 @@ def download_first_part(
     stop = False
 
     stats_collector = StatsCollector(response.url, f'{first_offset}:{last_offset}', 'hash')
-    stats_collector_read_append = stats_collector.read.append
-    stats_collector_other_append = stats_collector.other.append
-    stats_collector_write_append = stats_collector.write.append
-    start = before_read = perf_counter_ns()
-    for data in response.iter_content(chunk_size=chunk_size):
-        stats_collector_read_append(perf_counter_ns() - before_read)
-        if first_offset + bytes_read + len(data) >= last_offset:
-            to_write = data[:last_offset - bytes_read]
-            stop = True
-        else:
-            to_write = data
-        before_put = perf_counter_ns()
-        writer_queue_put((False, first_offset + bytes_read, to_write))
-
-        before_hash = perf_counter_ns()
-        hasher_update(to_write)
-        after_hash = perf_counter_ns()
-
-        stats_collector_write_append(before_hash - before_put)
-        stats_collector_other_append(after_hash - before_hash)
-
-        bytes_read += len(to_write)
-        if stop:
-            break
-
-    # since we got everything we need from original response, close the socket and free the buffer
-    # to avoid a timeout exception during hashing and other trouble
-    response.close()
-
-    url = response.request.url
-    tries_left = 5 - 1  # this is hardcoded because we are going to replace the entire retry interface soon, so we'll avoid deprecation here and keep it private
-    while tries_left and bytes_read < actual_part_size:
-        cloud_range = starting_cloud_range.subrange(
-            bytes_read, actual_part_size - 1
-        )  # first attempt was for the whole file, but retries are bound correctly
-        logger.debug(
-            'download attempts remaining: %i, bytes read already: %i. Getting range %s now.',
-            tries_left, bytes_read, cloud_range
-        )
-        with session.download_file_from_url(
-            url,
-            cloud_range.as_tuple(),
-            encryption=encryption,
-        ) as response:
-            before_read = perf_counter_ns()
-            for to_write in response.iter_content(chunk_size=chunk_size):
-                stats_collector_read_append(perf_counter_ns() - before_read)
 
-                before_put = perf_counter_ns()
+    with stats_collector.total:
+
+        stats_collector_read = stats_collector.read
+        stats_collector_other = stats_collector.other
+        stats_collector_write = stats_collector.write
```

Comment:
cosmetic: I'd move that to right after `stats_collector` is initialized, breaking the principle where we should declare variables as later as possible - I wouldn't want to muddy the code which actually does things with taking off the `.` to improve performance. As it is now, you are measuring those lines as part of "total" which may not skew the results too much, but still, it's like a lock, don't keep it longer than you need to.

Reply:
Done.